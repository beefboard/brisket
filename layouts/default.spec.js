import VueTestUtils from '@vue/test-utils'
import { shallowMount } from '@vue/test-utils'
import defaultLayout from './default'
import flushPromises from 'flush-promises'

VueTestUtils.config.stubs['nuxt-link'] = '<a><slot /></a>'
VueTestUtils.config.stubs['nuxt'] = '<div><slot /></div>'

describe('layout', () => {
  let mockAuth

  function renderLayout() {
    return shallowMount(defaultLayout, {
      mocks: {
        $store: {
          dispatch: jest.fn(),
          state: {
            auth: mockAuth
          }
        },
        $router: {
          push: jest.fn()
        }
      }
    })
  }

  beforeEach(() => {
    mockAuth = null
  })
  test('mounts properly', () => {
    const wrapper = renderLayout()
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  test('renders properly', () => {
    const wrapper = renderLayout()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should have a top bar', () => {
    const wrapper = renderLayout()
    const topbar = wrapper.find('.top-bar')

    expect(topbar.is('.top-bar')).toBe(true)
  })

  it('should only show link to login when not logged in', () => {
    let wrapper = renderLayout()
    let links = wrapper.find('.links')
    expect(links.find('[to="/login"]').exists()).toBe(true)

    mockAuth = {
      username: 'test'
    }
    wrapper = renderLayout()
    links = wrapper.find('.links')
    expect(links.find('[to="/login"]').exists()).toBe(false)
  })

  it('should only show link to register when not logged in', () => {
    let wrapper = renderLayout()
    let links = wrapper.find('.links')
    expect(links.find('[to="/register"]').exists()).toBe(true)

    mockAuth = {
      username: 'test'
    }
    wrapper = renderLayout()
    links = wrapper.find('.links')
    expect(links.find('[to="/register"]').exists()).toBe(false)
  })

  it('should only show link to logout when logged in', () => {
    let wrapper = renderLayout()
    let links = wrapper.find('.links')
    expect(links.find('[class="logout"]').exists()).toBe(false)

    mockAuth = {
      username: 'test'
    }

    wrapper = renderLayout()
    links = wrapper.find('.links')
    expect(links.find('[class="logout"]').exists()).toBe(true)
  })

  it('should only show link to profile when logged in', () => {
    let wrapper = renderLayout()
    let links = wrapper.find('.links')
    expect(links.find('[to="/profiles/test"]').exists()).toBe(false)

    mockAuth = {
      username: 'test'
    }

    wrapper = renderLayout()
    links = wrapper.find('.links')
    expect(links.find('[to="/profiles/test"]').exists()).toBe(true)
  })

  test('profile should link to profile for the username', () => {
    mockAuth = {
      username: 'test'
    }

    let wrapper = renderLayout()
    let links = wrapper.find('.links')
    expect(links.find('[to="/profiles/test"]').exists()).toBe(true)

    mockAuth = {
      username: 'test1'
    }

    wrapper = renderLayout()
    links = wrapper.find('.links')
    expect(links.find('[to="/profiles/test1"]').exists()).toBe(true)
  })

  test('logout link calls store logout', async () => {
    mockAuth = {
      username: 'test'
    }

    const wrapper = renderLayout()
    const links = wrapper.find('.links')
    const button = links.find('[class="logout"]')

    button.trigger('click')
    await flushPromises()

    expect(wrapper.vm.$store.dispatch).toHaveBeenCalledWith('logout')
  })

  test('logout link routes to / after logout', async () => {
    mockAuth = {
      username: 'test'
    }

    const wrapper = renderLayout()
    const links = wrapper.find('.links')
    const button = links.find('[class="logout"]')

    button.trigger('click')
    await flushPromises()

    expect(wrapper.vm.$router.push).toHaveBeenCalledWith('/')
  })
})
