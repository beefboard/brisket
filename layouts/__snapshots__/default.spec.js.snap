// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout renders properly 1`] = `
<div class="main-container">
  <!---->
  <div class="menu menu-closed">
    <div class="menu-profile-links"><a to="/login">Login</a> <a to="/register">Register</a>
      <!---->
      <!---->
    </div>
    <div class="menu-main-links"><a to="/" exact="">
        Home
      </a>
      <!---->
    </div>
  </div>
  <div class="top-bar">
    <div class="menu-button">
      <div icon="[object Object]" style="font-size: 2rem;"></div>
    </div> <a to="/">
      <div class="logo">
        <div class="img-wrapper"><img src="/img/meat.jpg"></div>
        <div class="logo-text">
          <div>Beefboard</div>
          <div class="sub-heading">Your local beef</div>
        </div>
      </div>
    </a>
    <div class="links">
      <div class="main-link"><a to="/" exact="">
          Home
        </a>
        <!---->
      </div>
      <div class="profile-links"><a to="/login">Login</a> <a to="/register">Register</a>
        <!---->
        <!---->
      </div>
    </div>
  </div>
  <div class="top-bar-spacer"></div>
  <div></div>
</div>
`;
